asset/dcr/addresses.go

```diff
@@ -10,68 +10,81 @@ import (
 )
 
 // AddressesByAccount handles a getaddressesbyaccount request by returning
-// all addresses for an account, or an error if the requested account does
-// not exist. Does not include the current address that can be retrieved with
-// w.mainWallet.CurrentAddress.
-func (w *Wallet) AddressesByAccount(ctx context.Context, account string) ([]string, error) {
+// external addresses for an account, or an error if the requested account does
+// not exist. Returns used and unused addresses up to nUsed and nUnused. No
+// unused addresses are returned if nUnused is zero. All used addresses are
+// returned if nUsed is zero. index is the first unused index.
+func (w *Wallet) AddressesByAccount(ctx context.Context, account string, nUsed, nUnused uint32) (used, unused []string, index uint32, err error) {
 	if account == "imported" {
 		addrs, err := w.mainWallet.ImportedAddresses(ctx, account)
 		if err != nil {
-			return nil, err
+			return nil, nil, 0, err
 		}
-		addrStrs := make([]string, len(addrs))
+		addrStrs := make([]string, 0, len(addrs))
 		for i := range addrs {
-			addrStrs[i] = addrs[i].String()
+			if nUsed != 0 && uint32(i) >= nUsed {
+				break
+			}
+			addrStrs = append(addrStrs, addrs[i].String())
 		}
-		return addrStrs, nil
+		return addrStrs, nil, 0, nil
 	}
 	accountNum, err := w.mainWallet.AccountNumber(ctx, account)
 	if err != nil {
-		return nil, err
+		return nil, nil, 0, err
 	}
 	xpub, err := w.mainWallet.AccountXpub(ctx, accountNum)
 	if err != nil {
-		return nil, err
+		return nil, nil, 0, err
 	}
 	extBranch, err := xpub.Child(0)
 	if err != nil {
-		return nil, err
-	}
-	intBranch, err := xpub.Child(1)
-	if err != nil {
-		return nil, err
+		return nil, nil, 0, err
 	}
-	endExt, endInt, err := w.mainWallet.BIP0044BranchNextIndexes(ctx, accountNum)
+	// It looks like this can never be zero, so a new wallet will always
+	// have a used address.
```

Comment:
i wouldn't bet on this always being the case

Reply:
Removing the comment.